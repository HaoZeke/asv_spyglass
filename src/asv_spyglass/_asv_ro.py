import json
import pprint as pp
import re
from pathlib import Path
from typing import Dict, Iterator, List, Union

from asv.util import load_json as asv_json_load

from asv_spyglass._aux import getstrform


class ReadOnlyASVBenchmarks:
    """Read-only holder for a set of ASV benchmarks."""

    api_version = 2

    def __init__(self, benchmarks_file: Path, regex: Union[str, List[str]] = None):
        """
        Initialize and load benchmarks from a JSON file, optionally filtering them.

        Args:
            benchmarks_file (Path): Path to the benchmarks JSON file.
            regex (Union[str, List[str]], optional): Regular expression(s) to filter benchmarks.
                Defaults to None (all benchmarks included).

        Raises:
            UserError: If the benchmarks file is missing or invalid.
        """
        d = asv_json_load(getstrform(benchmarks_file), api_version=self.api_version)
        self._benchmarks = d.values()
        self._filtered_benchmarks = self._filter_benchmarks(regex)

    def __iter__(self) -> Iterator[Dict]:
        """Iterate over the filtered benchmarks."""
        return iter(self._filtered_benchmarks.values())

    def __getitem__(self, name: str) -> Union[Dict, None]:
        """Get a benchmark by name."""
        return self._filtered_benchmarks.get(name)

    def __repr__(self) -> str:
        return pp.pformat(self._filtered_benchmarks)

    @property
    def benchmark_names(self) -> List[str]:
        """Get a list of benchmark names."""
        return list(self._filtered_benchmarks.keys())

    def _filter_benchmarks(
        self, regex: Union[str, List[str]] = None
    ) -> Dict[str, Dict]:
        """
        Filter benchmarks based on provided regular expression(s).

        Args:
            regex (Union[str, List[str]], optional): Regular expression(s) to filter benchmarks.
                Defaults to None (all benchmarks included).

        Returns:
            Dict[str, Dict]: A dictionary containing filtered benchmarks.
        """
        if not regex:
            return {benchmark["name"]: benchmark for benchmark in self._benchmarks}

        if isinstance(regex, str):
            regex = [regex]

        filtered_benchmarks = {}
        for benchmark in self._benchmarks:
            if any(re.search(r, benchmark["name"]) for r in regex):
                filtered_benchmarks[benchmark["name"]] = benchmark
        return filtered_benchmarks
