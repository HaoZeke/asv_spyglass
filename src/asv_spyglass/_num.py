import math
from dataclasses import dataclass


@dataclass
class BenchNum:
    val: float
    err: float | None
    unit: str | None


class Ratio:
    """
    Represents the ratio between two numeric values (t2 / t1), handling
    potential issues like division by zero, NaN, and None values. It also
    supports marking a ratio as insignificant.

    Attributes:
        _t1 (float | None): The first value (denominator).
        _t2 (float | None): The second value (numerator).
        val (float): The calculated ratio, which can be:
            - The actual ratio if both _t1 and _t2 are valid numbers and _t1 is not zero.
            - math.inf if either _t1 or _t2 is None or NaN, or if _t1 is zero.
        is_insignificant (bool): A flag indicating whether the ratio should be
            considered statistically insignificant. Defaults to False.

    Methods:
        __init__(self, t1: float | None, t2: float | None):
            Initializes a Ratio object, calculating the ratio if possible.

        __repr__(self):
            Returns a string representation of the ratio:
            - "n/a" if the ratio is undefined (val is math.inf).
            - "~" followed by the formatted ratio if is_insignificant is True.
            - A formatted string with 2 decimal places otherwise.

        _is_invalid(self, t1, t2):
            A private helper method to check if the inputs are invalid for ratio calculation.
    """

    def __init__(self, t1: float | None, t2: float | None):
        """
        Initializes a Ratio object.

        Args:
            t1 (float | None): The first value (denominator).
            t2 (float | None): The second value (numerator).
        """
        self._t1 = t1
        self._t2 = t2
        self.is_insignificant = False
        self.val = None

        if self._is_invalid(t1, t2):
            self.val = math.inf
        else:
            try:
                self.val = t2 / t1
            except ZeroDivisionError:
                self.val = math.inf

    def __repr__(self):
        """
        Returns a string representation of the ratio.

        If val is math.inf, returns "n/a".
        If is_insignificant is True, returns "~" followed by the formatted ratio.
        Otherwise, returns the ratio formatted to 2 decimal places.
        """
        if self.val == math.inf:
            return "n/a"
        elif self.is_insignificant:
            return "~" + f"{self.val:6.2f}".strip()
        else:
            return f"{self.val:6.2f}"

    def _is_invalid(self, t1, t2):
        """
        Checks if the inputs are invalid for ratio calculation.

        Args:
            t1 (float | None): The first value.
            t2 (float | None): The second value.

        Returns:
            bool: True if either t1 or t2 is None or NaN, False otherwise.
        """
        return t1 is None or t2 is None or math.isnan(t1) or math.isnan(t2)
