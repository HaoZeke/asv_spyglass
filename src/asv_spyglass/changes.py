import enum
from dataclasses import dataclass


class ResultColor(enum.StrEnum):
    DEFAULT = "black"
    GREEN = enum.auto()
    RED = enum.auto()
    LIGHTGREY = enum.auto()


class ResultMark(enum.StrEnum):
    BETTER = "-"
    WORSE = "+"
    FAILURE = "!"
    FIXED = "*"
    INCOMPARABLE = "x"
    UNCHANGED = " "
    INSIGNIFICANT = "~"


class AfterIs(enum.Enum):
    LUKEWARM = 0
    WORSE = -1
    BETTER = 1


@dataclass
class ASVChangeInfo:
    mark: ResultMark
    color: ResultColor
    description: str
    state: AfterIs
    before: str
    after: str


@dataclass
class Incomparable(ASVChangeInfo):
    mark: ResultMark = ResultMark.INCOMPARABLE
    color: ResultColor = ResultColor.LIGHTGREY
    description: str = "Not comparable"
    state: AfterIs = AfterIs.LUKEWARM
    before: str = ""
    after: str = ""


@dataclass
class Failure(ASVChangeInfo):
    mark: ResultMark = ResultMark.FAILURE
    color: ResultColor = ResultColor.RED
    description: str = "Introduced a failure"
    state: AfterIs = AfterIs.WORSE
    before: str = "Succeeded"
    after: str = "Failed"


@dataclass
class Fixed(ASVChangeInfo):
    mark: ResultMark = ResultMark.FIXED
    color: ResultColor = ResultColor.GREEN
    description: str = "Fixed a failure"
    state: AfterIs = AfterIs.BETTER
    before: str = "Failed"
    after: str = "Succeeded"


@dataclass
class NoChange(ASVChangeInfo):
    mark: ResultMark = ResultMark.UNCHANGED
    color: ResultColor = ResultColor.DEFAULT
    description: str = "Both failed or either was skipped or no significant change"
    state: AfterIs = AfterIs.LUKEWARM
    before: str = ""
    after: str = ""


@dataclass
class Better(ASVChangeInfo):
    mark: ResultMark = ResultMark.BETTER
    color: ResultColor = ResultColor.GREEN
    description: str = "Relative improvement"
    state: AfterIs = AfterIs.BETTER
    before: str = "Worse"
    after: str = "Better"


@dataclass
class Worse(ASVChangeInfo):
    mark: ResultMark = ResultMark.WORSE
    color: ResultColor = ResultColor.RED
    description: str = "Relatively worse"
    state: AfterIs = AfterIs.WORSE
    before: str = "Better"
    after: str = "Worse"
