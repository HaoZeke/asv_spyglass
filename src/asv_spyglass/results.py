import dataclasses
import math
import re
from collections import namedtuple
from dataclasses import dataclass, field

import polars as pl
from asv_runner.statistics import get_err

ASVResult = namedtuple(
    "ASVResult",
    [
        "key",
        "params",
        "value",
        "stats",
        "samples",
        "version",
        "machine",
        "env_name",
    ],
)


def result_iter(bdot):
    for key in bdot.get_all_result_keys():
        params = bdot.get_result_params(key)
        result_value = bdot.get_result_value(key, params)
        result_stats = bdot.get_result_stats(key, params)
        result_samples = bdot.get_result_samples(key, params)
        result_version = bdot.benchmark_version.get(key)
        yield ASVResult(
            key,
            params,
            result_value,
            result_stats,
            result_samples,
            result_version,
            bdot.params["machine"],
            bdot.env_name,
        )


@dataclass
class PreparedResult:
    """Augmented with information from the benchmarks.json"""

    units: dict
    results: dict
    stats: dict
    versions: dict
    machine_name: str
    env_name: str
    param_names: list

    def __iter__(self):
        for _field in dataclasses.fields(self):
            yield getattr(self, _field.name)

    def to_df(self):
        """
        Converts a PreparedResult object to a DataFrame, exploding the
        parameters in the keys and flattening the results.
        """

        data = []
        for key, result in self.results.items():
            # Extract benchmark name and parameters
            match = re.match(r"(.+)\((.*)\)", key)
            if match:
                benchmark_name = match.group(1)
                params = match.group(2).split(", ")
            else:
                benchmark_name = key
                params = []

            # Flatten the results tuple
            stats_dict, samples = self.stats[key]
            if stats_dict is None:
                stats_dict_maybe_null = {
                    "ci_99_a": None,
                    "ci_99_b": None,
                    "q_25": None,
                    "q_75": None,
                    "number": None,
                    "repeat": None,
                }
            else:
                stats_dict_maybe_null = stats_dict

            row = {
                "benchmark_base": benchmark_name,
                "name": key,
                "result": result,
                "units": self.units[key],
                "machine": self.machine_name,
                "env": self.env_name,
                "version": self.versions[key],
                **stats_dict_maybe_null,  # Expand the stats dictionary
                "samples": samples,
            }
            # Combine param and parameter name for column names
            row.update(
                dict(
                    zip(
                        [f"param_{name}" for name in self.param_names[key]],
                        params,
                    )
                )
            )
            data.append(row)

        return pl.DataFrame(data)


@dataclass
class ASVBench:
    name: str
    _pr: PreparedResult
    time: float = field(init=False)
    stats_n_samples: tuple = field(init=False)
    err: float | None = field(init=False)
    version: str | None = field(init=False)
    unit: str | None = field(init=False)

    def __post_init__(self):
        self.time = self._pr.results.get(self.name, math.nan)
        self.stats_n_samples = self._pr.stats.get(self.name, (None,))
        self.err = None
        if self.name in self._pr.stats and self.stats_n_samples[0]:
            self.err = get_err(self.time, self.stats_n_samples[0])
        self.version = self._pr.versions.get(self.name)
        self.unit = self._pr.units.get(self.name)
