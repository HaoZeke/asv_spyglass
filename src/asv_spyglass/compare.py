from pathlib import Path

from asv import results
from asv.commands.compare import _is_result_better, _isna, unroll_result
from asv.util import human_value
from asv_runner.console import color_print

from asv_spyglass._asv_ro import ReadOnlyASVBenchmarks
from asv_spyglass._num import Ratio
from asv_spyglass.changes import (
    ASVChangeInfo,
    Better,
    Failure,
    Fixed,
    Incomparable,
    NoChange,
    Worse,
)
from asv_spyglass.results import ASVBench, PreparedResult, result_iter


class ResultPreparer:
    """
    Prepares benchmark results for comparison by extracting relevant data
    like units, values, stats, versions, and machine/environment names.
    """

    def __init__(self, benchmarks):
        """
        Initializes ResultPreparer with benchmark data.

        Args:
            benchmarks: Benchmark data used for extracting units.
        """
        self.benchmarks = benchmarks

    def prepare(self, result_data):
        """
        Processes result data and returns extracted information.

        Args:
            result_data: Result data to be processed.

        Returns:
            tuple: A tuple containing units, results, stats, versions,
                   and the machine/environment name.
        """
        units = {}
        results = {}
        ss = {}
        versions = {}
        param_names = {}
        btype = {}
        machine_env_name = None

        for (
            key,
            params,
            value,
            stats,
            samples,
            version,
            machine,
            env_name,
        ) in result_iter(result_data):
            machine_env_name = f"{machine}/{env_name}"
            for name, value, stats, samples in unroll_result(
                key, params, value, stats, samples
            ):
                results[name] = value
                # TODO(hz): Split samples out later when _is_result_better is
                # not used anymore, currently in a tuple for ASV compatibility
                ss[name] = (stats, samples)
                versions[name] = version
                # HACK(hz): The names already include the parameters i.e.
                # benchmark(param) so this works around the situation
                bench_key = [x for x in self.benchmarks.keys() if key in x][0]
                units[name] = self.benchmarks.get(bench_key, {}).get("unit")
                btype[name] = self.benchmarks.get(bench_key, {}).get("type")
                param_names[name] = self.benchmarks.get(bench_key, {}).get(
                    "param_names"
                )

        machine, env_name = machine_env_name.split("/")
        return PreparedResult(
            units=units,
            results=results,
            stats=ss,
            versions=versions,
            machine_name=machine,
            env_name=env_name,
            param_names=param_names,
        )


def _get_change_info(
    asv1: ASVBench, asv2: ASVBench, factor, use_stats
) -> ASVChangeInfo:
    if (
        asv1.version is not None
        and asv2.version is not None
        and asv1.version != asv2.version
    ):
        # not comparable
        return Incomparable()
    elif asv1.time is not None and asv2.time is None:
        # introduced a failure
        return Failure()
    elif asv1.time is None and asv2.time is not None:
        # fixed a failure
        return Fixed()
    elif asv1.time is None and asv2.time is None:
        # both failed
        return NoChange()
    elif _isna(asv1.time) or _isna(asv2.time):
        # either one was skipped
        return NoChange()
    elif _is_result_better(
        asv2.time,
        asv1.time,
        asv2.stats_n_samples,
        asv1.stats_n_samples,
        factor,
        use_stats=use_stats,
    ):
        return Better()
    elif _is_result_better(
        asv1.time,
        asv2.time,
        asv1.stats_n_samples,
        asv2.stats_n_samples,
        factor,
        use_stats=use_stats,
    ):
        return Worse()
    else:
        return NoChange()


def do_compare(
    b1,
    b2,
    bdat,
    factor=1.1,
    split=False,
    only_changed=False,
    sort="default",
    machine=None,
    env_spec=None,
    use_stats=True,
):
    # Load results
    res_1 = results.Results.load(b1)
    res_2 = results.Results.load(b2)

    # Initialize benchmarks
    benchmarks = ReadOnlyASVBenchmarks(Path(bdat)).benchmarks

    # Prepare results using the ResultPreparer class
    preparer = ResultPreparer(benchmarks)
    pr1 = preparer.prepare(res_1)
    pr2 = preparer.prepare(res_2)

    # Extract data from prepared results
    machine_env_names = set()
    mname_1 = f"{pr1.machine_name}/{pr1.env_name}"
    mname_2 = f"{pr2.machine_name}/{pr2.env_name}"
    machine_env_names.add(mname_1)
    machine_env_names.add(mname_2)

    benchmarks_1 = set(pr1.results.keys())
    benchmarks_2 = set(pr2.results.keys())
    joint_benchmarks = sorted(list(benchmarks_1 | benchmarks_2))
    bench = {}

    if split:
        bench["green"] = []
        bench["red"] = []
        bench["lightgrey"] = []
        bench["default"] = []
    else:
        bench["all"] = []

    states = []
    for benchmark in joint_benchmarks:
        asv1 = ASVBench(benchmark, pr1)
        asv2 = ASVBench(benchmark, pr2)

        ratio = Ratio(asv1.time, asv2.time)
        diffinfo = _get_change_info(
            asv1,
            asv2,
            factor,
            use_stats,
        )
        states.append(diffinfo.state)

        if isinstance(diffinfo, NoChange):
            # Mark statistically insignificant results
            if _is_result_better(
                asv1.time, asv2.time, None, None, factor
            ) or _is_result_better(asv2.time, asv1.time, None, None, factor):
                ratio.is_insignificant = True

        if only_changed and diffinfo.mark.value in (" ", "x", "*"):
            continue

        details = (
            f"{diffinfo.mark:1s} "
            f"{human_value(asv1.time, asv1.unit, err=asv1.err):>15s} "
            f"{human_value(asv2.time, asv2.unit, err=asv2.err):>15s} "
            f"{str(ratio):>8s} "
        )
        split_line = details.split()
        if len(machine_env_names) > 1:
            benchmark_name = f"{benchmark} [{mname_1} -> {mname_2}]"
        else:
            benchmark_name = benchmark
        if len(split_line) == 4:
            split_line += [benchmark_name]
        else:
            split_line = [" "] + split_line + [benchmark_name]
        if split:
            bench[diffinfo.color].append(split_line)
        else:
            bench["all"].append(split_line)

    if split:
        keys = ["green", "default", "red", "lightgrey"]
    else:
        keys = ["all"]

    titles = {}
    titles["green"] = "Benchmarks that have improved:"
    titles["default"] = "Benchmarks that have stayed the same:"
    titles["red"] = "Benchmarks that have got worse:"
    titles["lightgrey"] = "Benchmarks that are not comparable:"
    titles["all"] = "All benchmarks:"

    all_tables = {}  # Dictionary to hold tables for each key

    for key in keys:
        if len(bench[key]) == 0:
            continue

        if not only_changed:
            color_print("")
            color_print(titles[key])
            color_print("")

        name_1 = False  # commit_names.get(hash_1)
        if name_1:
            name_1 = f"<{name_1}>"
        else:
            name_1 = ""

        name_2 = False  # commit_names.get(hash_2)
        if name_2:
            name_2 = f"<{name_2}>"
        else:
            name_2 = ""

        if sort == "default":
            pass
        elif sort == "ratio":
            bench[key].sort(key=lambda v: v[3], reverse=True)
        elif sort == "name":
            bench[key].sort(key=lambda v: v[2])
        else:
            raise ValueError("Unknown 'sort'")

        print(states)
        table_data = [
            [
                row[0],  # Change mark
                row[1],  # Before
                row[2],  # After
                row[3],  # Ratio
                row[4],  # Benchmark name
            ]
            for row in bench[key]
        ]

        all_tables[key] = {
            "title": titles[key],
            "headers": [
                "Change",
                f"Before {name_1}",
                f"After {name_2}",
                "Ratio",
                "Benchmark (Parameter)",
            ],
            "table_data": table_data,
            "states": states,
        }

    return all_tables
