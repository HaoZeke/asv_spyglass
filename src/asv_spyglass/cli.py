from pathlib import Path

import click
import polars as pl
from asv import results
from rich import box
from rich.console import Console
from rich.table import Table
from rich_click import RichCommand, RichGroup

from asv_spyglass._asv_ro import ReadOnlyASVBenchmarks
from asv_spyglass.changes import ResultMark
from asv_spyglass.compare import ResultPreparer, do_compare


@click.group(cls=RichGroup)
def cli():
    """
    Command-line interface for ASV benchmark analysis.
    """
    pass


@cli.command(cls=RichCommand)
@click.argument("b1", type=click.Path(exists=True), required=True)
@click.argument("b2", type=click.Path(exists=True), required=True)
@click.argument("bconf", type=click.Path(exists=True), required=True)
@click.option(
    "--factor",
    default=1.1,
    show_default=True,
    help="Factor for determining significant changes.",
)
@click.option(
    "--split",
    is_flag=True,
    help="Split output by improvement/regression.",
)
@click.option(
    "--only-changed",
    is_flag=True,
    help="Only show changed benchmarks.",
)
@click.option(
    "--sort",
    type=click.Choice(["default", "ratio", "name"]),
    default="default",
    show_default=True,
    help="Sort output by change, ratio, or name.",
)
def compare(b1, b2, bconf, factor, split, only_changed, sort):
    """
    Compare two ASV result files.
    """
    all_tables = do_compare(b1, b2, bconf, factor, split, only_changed, sort)

    console = Console()

    for key, table_data in all_tables.items():
        if not only_changed:
            console.print("")
            console.print(table_data["title"], style="bold")
            console.print("")

        table = Table(
            title=table_data["title"],
            show_header=True,
            header_style="bold magenta",
            box=box.SIMPLE,
        )

        for header in table_data["headers"]:
            table.add_column(
                header, justify="right" if header != "Benchmark (Parameter)" else "left"
            )

        for row in table_data["table_data"]:
            change_mark = row[0]
            row_style = ""

            # Determine row style based on change_mark
            if change_mark == ResultMark.BETTER:
                row_style = "green"
            elif change_mark == ResultMark.WORSE:
                row_style = "red"
            elif change_mark == ResultMark.FAILURE:
                row_style = "red"
            elif change_mark == ResultMark.FIXED:
                row_style = "green"
            elif change_mark == ResultMark.INCOMPARABLE:
                row_style = "light_grey"
            elif change_mark == ResultMark.UNCHANGED:
                row_style = "white"
            elif change_mark == ResultMark.INSIGNIFICANT:
                row_style = "white"

            table.add_row(*row, style=row_style)

        console.print(table)

        # Print summary of worsened/improved status
        if not split:
            if (av_x := sum([x.value for x in table_data["states"]])) > 0:
                console.print("[bold green]Net Improvement![/]")
            elif av_x < 0:
                console.print("[bold red]Net Regression![/]")


@cli.command(cls=RichCommand)
@click.argument("bres", type=click.Path(exists=True), required=True)
@click.argument("bdat", type=click.Path(exists=True), required=True)
@click.option(
    "--csv",
    type=click.Path(),
    help="Save data to csv",
)
def to_df(bres, bdat, csv):
    """
    Generate a dataframe from an ASV result file.
    """
    res = results.Results.load(bres)
    benchdat = ReadOnlyASVBenchmarks(Path(bdat)).benchmarks
    preparer = ResultPreparer(benchdat)
    df = preparer.prepare(res).to_df()
    if csv:
        df.to_csv(csv)
    else:
        with pl.Config(
            tbl_formatting="ASCII_MARKDOWN",
            tbl_hide_column_data_types=True,
            fmt_str_lengths=50,
            tbl_cols=50,
        ):
            click.echo(df)
