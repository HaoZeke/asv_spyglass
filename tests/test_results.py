from pathlib import Path

from approvaltests.approvals import verify
import pprint as pp

from asv import results
from asv_spyglass.compare import result_iter


# Kanged from rgpycrumbs
def getstrform(pathobj):
    return str(pathobj.absolute())


def test_result_iter(shared_datadir):
    bdot = results.Results.load(
        getstrform(shared_datadir / "a0f29428-conda-py3.11-numpy.json")
    )
    verify(pp.pformat(list(result_iter(bdot))))
