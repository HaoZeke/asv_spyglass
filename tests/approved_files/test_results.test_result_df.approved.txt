{'benchmark_base': {0: 'benchmarks.TimeSuiteDecoratorSingle.time_keys',
                    1: 'benchmarks.TimeSuiteDecoratorSingle.time_keys',
                    2: 'benchmarks.TimeSuiteDecoratorSingle.time_keys',
                    3: 'benchmarks.TimeSuiteDecoratorSingle.time_values',
                    4: 'benchmarks.TimeSuiteDecoratorSingle.time_values',
                    5: 'benchmarks.TimeSuiteDecoratorSingle.time_values',
                    6: 'benchmarks.TimeSuiteMultiDecorator.time_ranges',
                    7: 'benchmarks.TimeSuiteMultiDecorator.time_ranges',
                    8: 'benchmarks.TimeSuiteMultiDecorator.time_ranges',
                    9: 'benchmarks.TimeSuiteMultiDecorator.time_ranges',
                    10: 'benchmarks.time_ranges_multi',
                    11: 'benchmarks.time_ranges_multi',
                    12: 'benchmarks.time_ranges_multi',
                    13: 'benchmarks.time_ranges_multi',
                    14: 'benchmarks.time_sort',
                    15: 'benchmarks.time_sort'},
 'ci_99_a': {0: 1.3444e-07,
             1: 5.3813e-07,
             2: 1.0672e-06,
             3: 1.8023e-07,
             4: 7.7445e-07,
             5: 1.4525e-06,
             6: 2.2972e-07,
             7: 1.366e-06,
             8: 6.4741e-07,
             9: 3.7563e-06,
             10: 1.7102e-07,
             11: 1.0348e-06,
             12: 4.3157e-07,
             13: 3.0563e-06,
             14: 1.0645e-06,
             15: 1.6189e-06},
 'ci_99_b': {0: 1.4947e-07,
             1: 5.4586e-07,
             2: 1.091e-06,
             3: 1.9282e-07,
             4: 7.9307e-07,
             5: 1.4623e-06,
             6: 2.3326e-07,
             7: 1.3947e-06,
             8: 6.5358e-07,
             9: 3.9037e-06,
             10: 1.7728e-07,
             11: 1.0653e-06,
             12: 4.3557e-07,
             13: 3.1576e-06,
             14: 1.3746e-06,
             15: 1.6341e-06},
 'env': {0: 'rattler-py3.12-numpy',
         1: 'rattler-py3.12-numpy',
         2: 'rattler-py3.12-numpy',
         3: 'rattler-py3.12-numpy',
         4: 'rattler-py3.12-numpy',
         5: 'rattler-py3.12-numpy',
         6: 'rattler-py3.12-numpy',
         7: 'rattler-py3.12-numpy',
         8: 'rattler-py3.12-numpy',
         9: 'rattler-py3.12-numpy',
         10: 'rattler-py3.12-numpy',
         11: 'rattler-py3.12-numpy',
         12: 'rattler-py3.12-numpy',
         13: 'rattler-py3.12-numpy',
         14: 'rattler-py3.12-numpy',
         15: 'rattler-py3.12-numpy'},
 'machine': {0: 'rgx1gen11',
             1: 'rgx1gen11',
             2: 'rgx1gen11',
             3: 'rgx1gen11',
             4: 'rgx1gen11',
             5: 'rgx1gen11',
             6: 'rgx1gen11',
             7: 'rgx1gen11',
             8: 'rgx1gen11',
             9: 'rgx1gen11',
             10: 'rgx1gen11',
             11: 'rgx1gen11',
             12: 'rgx1gen11',
             13: 'rgx1gen11',
             14: 'rgx1gen11',
             15: 'rgx1gen11'},
 'name': {0: 'benchmarks.TimeSuiteDecoratorSingle.time_keys(10)',
          1: 'benchmarks.TimeSuiteDecoratorSingle.time_keys(100)',
          2: 'benchmarks.TimeSuiteDecoratorSingle.time_keys(200)',
          3: 'benchmarks.TimeSuiteDecoratorSingle.time_values(10)',
          4: 'benchmarks.TimeSuiteDecoratorSingle.time_values(100)',
          5: 'benchmarks.TimeSuiteDecoratorSingle.time_values(200)',
          6: "benchmarks.TimeSuiteMultiDecorator.time_ranges(10, 'range')",
          7: "benchmarks.TimeSuiteMultiDecorator.time_ranges(10, 'arange')",
          8: "benchmarks.TimeSuiteMultiDecorator.time_ranges(100, 'range')",
          9: "benchmarks.TimeSuiteMultiDecorator.time_ranges(100, 'arange')",
          10: "benchmarks.time_ranges_multi(10, 'range')",
          11: "benchmarks.time_ranges_multi(10, 'arange')",
          12: "benchmarks.time_ranges_multi(100, 'range')",
          13: "benchmarks.time_ranges_multi(100, 'arange')",
          14: 'benchmarks.time_sort(10)',
          15: 'benchmarks.time_sort(100)'},
 'number': {0: 67364,
            1: 16815,
            2: 8960,
            3: 63961,
            4: 15516,
            5: 8748,
            6: 51698,
            7: 9417,
            8: 20005,
            9: 3213,
            10: 54349,
            11: 9631,
            12: 20588,
            13: 3042,
            14: 9345,
            15: 5828},
 'param_0_n': {0: nan,
               1: nan,
               2: nan,
               3: nan,
               4: nan,
               5: nan,
               6: '10',
               7: '10',
               8: '100',
               9: '100',
               10: '10',
               11: '10',
               12: '100',
               13: '100',
               14: '10',
               15: '100'},
 'param_0_size': {0: '10',
                  1: '100',
                  2: '200',
                  3: '10',
                  4: '100',
                  5: '200',
                  6: nan,
                  7: nan,
                  8: nan,
                  9: nan,
                  10: nan,
                  11: nan,
                  12: nan,
                  13: nan,
                  14: nan,
                  15: nan},
 'param_1_func_name': {0: nan,
                       1: nan,
                       2: nan,
                       3: nan,
                       4: nan,
                       5: nan,
                       6: "'range'",
                       7: "'arange'",
                       8: "'range'",
                       9: "'arange'",
                       10: "'range'",
                       11: "'arange'",
                       12: "'range'",
                       13: "'arange'",
                       14: nan,
                       15: nan},
 'q_25': {0: 1.3621e-07,
          1: 5.419e-07,
          2: 1.0707e-06,
          3: 1.8595e-07,
          4: 7.8003e-07,
          5: 1.4593e-06,
          6: 2.3038e-07,
          7: 1.3691e-06,
          8: 6.4921e-07,
          9: 3.8065e-06,
          10: 1.7412e-07,
          11: 1.0378e-06,
          12: 4.3176e-07,
          13: 3.0622e-06,
          14: 1.0715e-06,
          15: 1.6225e-06},
 'q_75': {0: 1.431e-07,
          1: 5.4495e-07,
          2: 1.0782e-06,
          3: 1.9121e-07,
          4: 7.8758e-07,
          5: 1.4621e-06,
          6: 2.3224e-07,
          7: 1.3791e-06,
          8: 6.5143e-07,
          9: 3.8388e-06,
          10: 1.7638e-07,
          11: 1.044e-06,
          12: 4.342e-07,
          13: 3.1181e-06,
          14: 1.0777e-06,
          15: 1.6326e-06},
 'repeat': {0: 10,
            1: 10,
            2: 10,
            3: 10,
            4: 10,
            5: 10,
            6: 10,
            7: 10,
            8: 10,
            9: 10,
            10: 10,
            11: 10,
            12: 10,
            13: 10,
            14: 10,
            15: 10},
 'result': {0: 1.37378532884605e-07,
            1: 5.429162653926854e-07,
            2: 1.0728282369663897e-06,
            3: 1.870523832540821e-07,
            4: 7.847095893996428e-07,
            5: 1.4613044126333168e-06,
            6: 2.3096740716340811e-07,
            7: 1.3737530519982098e-06,
            8: 6.507371149493031e-07,
            9: 3.831307194414474e-06,
            10: 1.75908075758346e-07,
            11: 1.04194673574825e-06,
            12: 4.322195698640758e-07,
            13: 3.0868006284377445e-06,
            14: 1.0754531829770897e-06,
            15: 1.6311290342036977e-06},
 'samples': {0: None,
             1: None,
             2: None,
             3: None,
             4: None,
             5: None,
             6: None,
             7: None,
             8: None,
             9: None,
             10: None,
             11: None,
             12: None,
             13: None,
             14: None,
             15: None},
 'units': {0: 'seconds',
           1: 'seconds',
           2: 'seconds',
           3: 'seconds',
           4: 'seconds',
           5: 'seconds',
           6: 'seconds',
           7: 'seconds',
           8: 'seconds',
           9: 'seconds',
           10: 'seconds',
           11: 'seconds',
           12: 'seconds',
           13: 'seconds',
           14: 'seconds',
           15: 'seconds'},
 'version': {0: '64746c9051ff76aa879b428c27b4247e8ed976c44a40579ae9aa4601b9883ed5',
             1: '64746c9051ff76aa879b428c27b4247e8ed976c44a40579ae9aa4601b9883ed5',
             2: '64746c9051ff76aa879b428c27b4247e8ed976c44a40579ae9aa4601b9883ed5',
             3: 'ab162b6142a1390a0e2a667ed8d2d3285f77152d9caa559b4fd784bb120aa153',
             4: 'ab162b6142a1390a0e2a667ed8d2d3285f77152d9caa559b4fd784bb120aa153',
             5: 'ab162b6142a1390a0e2a667ed8d2d3285f77152d9caa559b4fd784bb120aa153',
             6: 'bf8381bfa42fb47f084462e83ab870e9779fb25535168f77415db72845b02f6d',
             7: 'bf8381bfa42fb47f084462e83ab870e9779fb25535168f77415db72845b02f6d',
             8: 'bf8381bfa42fb47f084462e83ab870e9779fb25535168f77415db72845b02f6d',
             9: 'bf8381bfa42fb47f084462e83ab870e9779fb25535168f77415db72845b02f6d',
             10: 'f9ae8b134446c273c0d3eb1e90246ae0d6f99389d06119dfe4fed7210eccdb15',
             11: 'f9ae8b134446c273c0d3eb1e90246ae0d6f99389d06119dfe4fed7210eccdb15',
             12: 'f9ae8b134446c273c0d3eb1e90246ae0d6f99389d06119dfe4fed7210eccdb15',
             13: 'f9ae8b134446c273c0d3eb1e90246ae0d6f99389d06119dfe4fed7210eccdb15',
             14: '60785bf757da0254d857b696482db7a25509a5b28a2c9d2a54e85d2a593ce272',
             15: '60785bf757da0254d857b696482db7a25509a5b28a2c9d2a54e85d2a593ce272'}}
